File: MontecarloSearch.cs
using System;
using System.Collections.Generic;
using System.Linq;

// ディレクテッドエボリューションの拡張シミュレーション
// アナロジー：鍵（シーケンス）を変形して錠（ターゲット）に最適化
// 論理：点変異→適応度評価→選択を繰り返し、多様性を維持
class DirectedEvolution
{
    // パラメータ
    private const int PopulationSize = 100;       // 集団サイズ
    private const int MaxGenerations = 50;        // 最大世代数
    private const int SequenceLength = 10;        // シーケンス長
    private const double FitnessThreshold = 0.9;  // 適応度閾値
    private const double MutationRate = 1.0;      // 変異率（1シーケンスあたり1箇所）
    private static readonly char[] AminoAcids = { 'A', 'C', 'G', 'T' }; // アミノ酸
    private static readonly char[] Target = "AAAAAACCCC".ToCharArray(); // ターゲット
    private static readonly Random Random = new Random();

    // シーケンスクラス
    private class Sequence
    {
        public char[] AminoAcids { get; }
        public double Fitness { get; set; }

        public Sequence(char[] aminoAcids)
        {
            AminoAcids = aminoAcids;
        }

        // コピーコンストラクタ
        public Sequence(Sequence other) : this((char[])other.AminoAcids.Clone()) { }

        // 適応度計算：ハミング距離の逆数
        public void ComputeFitness(char[] target)
        {
            int hammingDistance = AminoAcids.Zip(target, (a, b) => a != b ? 1 : 0).Sum();
            Fitness = 1.0 / (1.0 + hammingDistance);
        }
    }

    // 多様性計算：集団内の平均ハミング距離
    private static double ComputeDiversity(List<Sequence> population)
    {
        double hammingSum = 0.0;
        double pairs = 0.0;
        for (int i = 0; i < population.Count - 1; i++)
        {
            for (int j = i + 1; j < population.Count; j++)
            {
                hammingSum += population[i].AminoAcids.Zip(population[j].AminoAcids, (a, b) => a != b ? 1 : 0).Sum();
                pairs += 1.0;
            }
        }
        return hammingSum / pairs / SequenceLength;
    }

    // ランダムシーケンス生成
    private static Sequence GenerateRandomSequence()
    {
        char[] aminoAcids = new char[SequenceLength];
        for (int i = 0; i < SequenceLength; i++)
        {
            aminoAcids[i] = AminoAcids[Random.Next(AminoAcids.Length)];
        }
        return new Sequence(aminoAcids);
    }

    // 変異：1箇所の点変異
    private static Sequence Mutate(Sequence seq)
    {
        if (Random.NextDouble() >= MutationRate) return new Sequence(seq);
        var mutated = new Sequence(seq);
        int pos = Random.Next(SequenceLength);
        mutated.AminoAcids[pos] = AminoAcids[Random.Next(AminoAcids.Length)];
        return mutated;
    }

    // メインシミュレーション
    public static void Run()
    {
        // 初期化：ランダムなシーケンス集団
        var population = Enumerable.Range(0, PopulationSize)
            .Select(_ => GenerateRandomSequence())
            .ToList();

        // 適応度履歴
        var fitnessHistory = new List<(double MaxFitness, double AvgFitness)>();

        // メインループ
        for (int gen = 1; gen <= MaxGenerations; gen++)
        {
            // 変異
            var mutants = population.Select(Mutate).ToList();

            // 評価：適応度
            foreach (var seq in mutants)
            {
                seq.ComputeFitness(Target);
            }

            // 多様性計算
            double diversity = ComputeDiversity(mutants);

            // 選択：上位50%を保持、残りはランダム生成
            var sorted = mutants.OrderByDescending(s => s.Fitness).ToList();
            population = sorted.Take(PopulationSize / 2)
                .Concat(Enumerable.Range(0, PopulationSize / 2).Select(_ => GenerateRandomSequence()))
                .ToList();

            // 適応度履歴
            double maxFitness = sorted[0].Fitness;
            double avgFitness = mutants.Average(s => s.Fitness);
            fitnessHistory.Add((maxFitness, avgFitness));

            // 結果表示
            Console.WriteLine($"Gen {gen:D3}: Best fitness = {maxFitness:F4}, Avg fitness = {avgFitness:F4}, Diversity = {diversity:F4}");
            Console.WriteLine($"Best seq = {new string(sorted[0].AminoAcids)}");

            // 終了条件
            if (maxFitness >= FitnessThreshold)
            {
                Console.WriteLine("Target fitness reached. Stopping.");
                break;
            }
        }

        // 最終結果：適応度履歴の要約
        Console.WriteLine("\nEvolution Summary:");
        Console.WriteLine("Generation | Max Fitness | Avg Fitness");
        for (int i = 0; i < fitnessHistory.Count; i++)
        {
            Console.WriteLine($"{i + 1,10} {fitnessHistory[i].MaxFitness,12:F4} {fitnessHistory[i].AvgFitness,12:F4}");
        }
    }

    // エントリーポイント
    static void Main()
    {
        Run();
    }
}

File: MontecarloSearch.f90
! ディレクテッドエボリューションの拡張シミュレーション
! アナロジー：鍵（シーケンス）を変形して錠（ターゲット）に最適化
! 論理：点変異→適応度評価→選択を繰り返し、多様性を維持
program directed_evolution
    implicit none
    ! パラメータ
    integer, parameter :: n_pop = 100        ! 集団サイズ
    integer, parameter :: n_gen = 50         ! 最大世代数
    integer, parameter :: seq_len = 10       ! シーケンス長
    real, parameter :: fitness_threshold = 0.9 ! 適応度閾値
    real, parameter :: mutation_rate = 1.0   ! 変異率（1シーケンスあたり1箇所）
    character :: amino_acids(4) = ['A', 'C', 'G', 'T'] ! アミノ酸（簡略化）
    character :: target(seq_len)             ! ターゲットシーケンス
    character :: pop(n_pop, seq_len)         ! シーケンス集団
    character :: mutants(n_pop, seq_len)     ! 変異体
    real :: fitnesses(n_pop)                 ! 適応度
    real :: fitness_history(n_gen, 2)        ! 最大・平均適応度の履歴
    integer :: sorted_idx(n_pop)             ! ソートインデックス
    integer :: gen, i, pos, aa_idx
    real :: r, avg_diversity

    ! ターゲットシーケンス：AAAAAACCCC
    target(1:6) = 'A'
    target(7:10) = 'C'

    ! 初期化：ランダムなシーケンス集団
    call random_seed()
    do i = 1, n_pop
        do pos = 1, seq_len
            call random_number(r)
            pop(i, pos) = amino_acids(int(r * 4) + 1)
        end do
    end do

    ! メインループ
    do gen = 1, n_gen
        ! 変異：1箇所の点変異
        mutants = pop
        do i = 1, n_pop
            call random_number(r)
            if (r < mutation_rate) then
                call random_number(r)
                pos = int(r * seq_len) + 1
                call random_number(r)
                aa_idx = int(r * 4) + 1
                mutants(i, pos) = amino_acids(aa_idx)
            end if
        end do

        ! 評価：適応度（ハミング距離の逆数）
        do i = 1, n_pop
            fitnesses(i) = compute_fitness(mutants(i,:), target)
        end do

        ! 多様性計算：集団の平均ハミング距離
        avg_diversity = compute_diversity(mutants, n_pop, seq_len)

        ! 選択：上位50%を保持、残りはランダム生成
        call sort(fitnesses, sorted_idx, n_pop)
        do i = 1, n_pop/2
            pop(i,:) = mutants(sorted_idx(i),:)
        end do
        do i = n_pop/2 + 1, n_pop
            do pos = 1, seq_len
                call random_number(r)
                pop(i, pos) = amino_acids(int(r * 4) + 1)
            end do
        end do

        ! 適応度履歴の記録
        fitness_history(gen, 1) = maxval(fitnesses) ! 最大適応度
        fitness_history(gen, 2) = sum(fitnesses) / n_pop ! 平均適応度

        ! 結果表示
        print '(A,I3,A,F6.4,A,F6.4,A,F6.4)', &
              'Gen ', gen, ': Best fitness = ', fitness_history(gen, 1), &
              ', Avg fitness = ', fitness_history(gen, 2), &
              ', Diversity = ', avg_diversity
        print '(A,10A1)', 'Best seq = ', (mutants(sorted_idx(1),i),i=1,seq_len)

        ! 終了条件：適応度閾値に達した場合
        if (fitness_history(gen, 1) >= fitness_threshold) then
            print *, 'Target fitness reached. Stopping.'
            exit
        end if
    end do

    ! 最終結果：適応度履歴の要約
    print *, 'Evolution Summary:'
    print *, 'Generation | Max Fitness | Avg Fitness'
    do gen = 1, min(n_gen, gen) ! 最後の世代まで
        print '(I10,2F12.4)', gen, fitness_history(gen, 1), fitness_history(gen, 2)
    end do

contains
    ! 適応度関数：ハミング距離の逆数
    real function compute_fitness(seq, target)
        character, intent(in) :: seq(seq_len), target(seq_len)
        integer :: hamming_distance
        hamming_distance = sum(merge(1, 0, seq /= target))
        compute_fitness = 1.0 / (1.0 + real(hamming_distance))
    end function compute_fitness

    ! 多様性計算：集団内の平均ハミング距離
    real function compute_diversity(pop, n_pop, seq_len)
        character, intent(in) :: pop(n_pop, seq_len)
        integer, intent(in) :: n_pop, seq_len
        integer :: i, j, hamming_sum
        real :: pairs
        hamming_sum = 0
        pairs = 0.0
        do i = 1, n_pop-1
            do j = i+1, n_pop
                hamming_sum = hamming_sum + sum(merge(1, 0, pop(i,:) /= pop(j,:)))
                pairs = pairs + 1.0
            end do
        end do
        compute_diversity = real(hamming_sum) / pairs / real(seq_len)
    end function compute_diversity

    ! ソート（簡略化：実際はバブルソートなど実装）
    subroutine sort(arr, idx, n)
        real, intent(in) :: arr(n)
        integer, intent(out) :: idx(n)
        integer, intent(in) :: n
        real :: temp_arr(n)
        integer :: i, j, temp_idx
        temp_arr = arr
        do i = 1, n
            idx(i) = i
        end do
        ! バブルソート（降順）
        do i = 1, n-1
            do j = 1, n-i
                if (temp_arr(j) < temp_arr(j+1)) then
                    temp_arr(j:j+1) = temp_arr(j+1:j:-1)
                    temp_idx = idx(j)
                    idx(j) = idx(j+1)
                    idx(j+1) = temp_idx
                end if
            end do
        end do
    end subroutine sort
end program directed_evolution
